JavaScript

gr2m opened the pull request:
This evolution proposal is part of point 3 of the Hubot 3.0 milestone:

> Modernize the project by translating CoffeeScript to JavaScript, improving integration with various developer tools, and adding features that make it easier for developers to automate their workflow.

**[Read the proposal](https://github.com/hubotio/evolution/blob/4/javascript/_drafts/javascript.md)** 👀

### Open todos / follow ups

- [x] Update [CONTRIBUTING.md](https://github.com/github/hubot/blob/master/CONTRIBUTING.md), remove mention of CoffeeScript (https://github.com/github/hubot/pull/1349)
- [x] Check in with https://github.com/github/hubot/issues/1338
- [x] add an automated end-to-end test for hubot scripts written in CoffeeScript (https://github.com/hubotio/hubot/pull/1360)
- [x] add tests for `require('hubot/es2015.js')` (https://github.com/hubotio/hubot/pull/1359)
- [ ] rename `src/` to `lib/` when everything else is done

gr2m committed:
Start proposal for CoffeeScript to JavaScript conversion

gr2m committed:
bullet points styling

gr2m committed:
style: bold

bkeepers commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,110 @@
+- [ ] Convert test files from CoffeeScript to JavaScript.
+- [ ] update package.json
+- [ ] add script for JavaScript linting
+- [ ] Update documentation
```
Everything in https://github.com/github/hubot/issues/1327 should also be in this list, especially https://github.com/github/generator-hubot

gr2m committed:
removed double mention of standardjs

mose commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,106 @@
+### Convert source files from CoffeeScript to JavaScript
+
+Convert all source files from CoffeeScript to JavaScript with a tool like [espresso](https://github.com/HipsterBrown/espresso).
+
```
The author of `decaffeinate` uses hubot code base for his tests of achievement, it's worth citing https://github.com/decaffeinate/decaffeinate#status and the https://github.com/github/hubot/issues/1138 issue.

gr2m replied:
good timing, I just found out about it and am looking into it right now. It looks like the better option for us. I’ll keep you posted

mose commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,106 @@
+
+  The Node Foundation has tried to get Debian to adapt modern Node versions like "[pretty much every other repository for packages takes modern versions.](https://twitter.com/mikeal/status/869646796888330240)" but could not succeed.
+
+  See more responses to my question regarding old Node versions at https://twitter.com/gr2m/status/869305267464306689
```
Debian users that need modern node version have 2 main easy options: 

- turning to [nvm](https://github.com/creationix/nvm)
- getting packages from [nodesource](https://github.com/nodesource/distributions)

gr2m replied:
good point, we could document these options as part of our breaking changes documentation / changelogs

gr2m committed:
decaffeinate is a better suited tool helping with the CoffeeScript to JavaScript conversion

cc @mose

gr2m committed:
add examples

gr2m committed:
add note that Hubot core does not use any promises at this point, hence async/await support would have no immediate effect

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+## Motivation
+
+This evolution proposal is part of point 3 of the Hubot 3.0 milestone:
```
Could you link to this?

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+## Detailed process
+
+The steps of the conversation will be
```
Is the intent to do each of these as a PR per item? One big PRs? Some combination?

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+Make sure the existing tests (still written in CoffeeScript) run against the new JavaScript.
+
+Now go trough each file and improve the code readability by hand as needed. From my experiences with converting Hoodie from CoffeeScript to JavaScript quite a lot of manual work will be required.
```
s/trough/through/

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+Make sure the existing tests (still written in CoffeeScript) run against the new JavaScript.
+
+Now go trough each file and improve the code readability by hand as needed. From my experiences with converting Hoodie from CoffeeScript to JavaScript quite a lot of manual work will be required.
```
To be clear, this is improving internal readability, and is leaving the external API in-tact?

gr2m replied:
correct

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+Same as the previous steps only for the tests files this time.
+
+When finished, create a separate branch with with the tests written in JavaScript but the source code still in CoffeeScript to assure integrity.
```
😍 

This sounds like a great way to build confidence in the conversion.

tombell replied:
There is a `with with` duplication in that paragraph.

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+### Update package.json
+
+Remove all CoffeeScript related dependencies and tooling. Update [engines](https://docs.npmjs.com/files/package.json#engines)
```
Do existing bots have their own coffeescript dependency, or do they rely on hubot's?

mose replied:
It would make sense to begin to build some script to make an inventory, like we had with https://github.com/hubotio/hubot-scripts-list so we can have data on existing population of plugins. I think I will give that a try,

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+Remove all CoffeeScript related dependencies and tooling. Update [engines](https://docs.npmjs.com/files/package.json#engines)
+
+Make sure that Hubot scripts can still be written in CoffeeScript. We want to break as few existing scripts as possible and many of them are written in CoffeeScript. We treat it as legacy support and will drop support for scripts written in CoffeeScript in future.
```
How can we make sure coffeescript tests still work? Maybe something like a test hubot instance for automated testing?

gr2m replied:
Good idea, I thought of testing it manually, but will try to add an automated e2e test for CI

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+### Update documentation
+
+Remove any reference of CoffeeScript from the documentation.
```
If it's considered legacy, maybe we should still have documentation until it's removed? Even if it's removed, it might be worth having some way to get to the old documentation (ie a link in the readme to an older sha?)

mose replied:
I feel that we need to duplicate the documentation, keep the old one somewhere. Because many people, for legitimate reason, will wait hubot3 stability before jumping in.

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+## Backward compatibility
+
+Only breaking change is dropping support of Node 0.10 and 0.12.
```
The way that hubot is launched could be a possible breakage:

- generated `bin/hubot` calls `node_modules/.bin/hubot`
- `node_modules/.bin/hubot` has a `#!/usr/bin/env coffee`
- that means it should be possible to just change the #! and have things work in github/hubot

It should be noted it's plausible that code in the wild calls `coffee node_modules/.bin/hubot` in the wild. Our internal hubot has a case where it does that for debugging. The failure case for that is probably syntax errors.

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+### Add script for JavaScript linting
+
+The only tool I would like to introduce as part of this proposal is a linting tool which will be run as part of `npm test`.
```
Could we make sure to update (or create?) contributor documentation around using the linter? I've had times where I've started on new process, with CI running linter, and the feedback process is relatively slow until you actually start getting it running locally, if not as part of your editor setup.

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+The only tool I would like to introduce as part of this proposal is a linting tool which will be run as part of `npm test`.
+
+I suggest [standard](https://standardjs.com/). We’ve been using it in all our projects at Hoodie and Neighbourhoodie since 2+ years and never looked back. It's a zero-configuration JavaScript linter.
```
👍 from me for standard. I've only used it a couple times, but I like that it is zero config, and the style it encourages is pretty light on line-noise (ie no semi-colons).

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+  Node versions smaller than v4 are no longer maintained and have known security vulnerabilities. The only major operating system which distributes a no longer maintained Node version by default is Debian, and there are alternative ways to install a recent Node version, see [Debian and Ubuntu based Linux distributions](https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux- ...
+
+  Mikeal Rogers (Node Foundation) says about Debian’s official Node packages:
```
Is there a relevant link for this, ie tweet, documentation, etc?

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+  More importantly, we would make it much harder for existing users to upgrade to the new Hubot version. Hubot is a widely used project and having a clear upgrade plan with a reasonable pace is critical. If we only drop support for versions which are no longer maintained we can align with Node’s [LTS schedule](https://github.com/nodejs/LTS#readme) so we can avoid a recurring discussion of when to ...
+
+  Besides, the current Hubot does not use a single promise in its core implementation at this point.
```
cc https://github.com/github/hubot/issues/1338 which is talking about adding it

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,294 @@
+
+  Node 7.6 introduced native support for async/await. For asynchronous code like much of Hubot is, this is a very nice feature. It makes the code more readable and hence more accessible for contributors which is a main objective of the Hubot community.
+
+  The reason we decided against it is that the complexity of Hubot core is not too high, the code can be made very readable even when using Promises. Once the one-time conversion from CoffeeScript to JavaScript is done, only smaller parts of the code base will be touched, the effect of async/await vs. promises will be limited.
```
I really like this sentiment.

I'm not familiar enough with async/await yet, but would it possible to still take advantage of these support in the core? That is, can hubot users use it even if the core wasn't specifically written for it?

gr2m replied:
Yes, Hubot user can take full advantage for the Node version they run their own Hubot instance in. If Hubot core adds Promise-based APIs then users which run their Hubot in Node 7.6+ can use async/await as these are sugar APIs to work with Promises. At the same time the core would continue to work with Node v4, so we don’t break Hubot instances that require this older Node version. Does that answer your question?

technicalpickles reviewed (changes requested):
This is a great writeup, thanks for putting it together!

I have tons of questions and feedback 😁 It is mostly from experience with working on hubot, and also to make this specific enough that it can be implemented by someone that isn't @gr2m

gr2m committed:
_drafts/

remove obsolete code in example

gr2m committed:
typo

gr2m commented:
I’m running into a bit of a blocker: when using the `class` keyword in the JavaScript version of Hubot, the way existing adapters inhert from `Adapter` would break because they are doing this

```js
MyAdapter.__super__.constructor.apply(this, arguments);
```

which fails with `TypeError: Class constructor Adapter cannot be invoked without 'new'`. There is a related issue here. Slack also [extends the `Message` class](https://github.com/slackapi/hubot-slack/blob/master/src/reaction-message.coffee), probably others do so, too.

I’m still thinking about how to find the best balance between readable, easy to contribute/maintain code on the one and as few breaking changes on the other side.

Let me know if you have any thoughts on this :)

technicalpickles commented:
> which fails with TypeError: Class constructor Adapter cannot be invoked without 'new'. There is a related issue here. Slack also extends the Message class, probably others do so, too.

I did a search for this error (ie `javascript class constructor cannot be invoked without new`), and seeing it come up quite a bit. There's a lot of github issues around specific libraries/tools. I think http://raganwald.com/2014/07/09/javascript-constructor-problem.html looks like a decent description of the problem with some proposed solutions.

> I’m still thinking about how to find the best balance between readable, easy to contribute/maintain code on the one and as few breaking changes on the other side.

Initially, I'd be most concerned with interoperability with existing code, unless we are committing to updating all the existing adapters as part of this work. I suspect less readable, harder to contribute to code can be confined to specific parts of the code base. Combine generous amounts of code comments about what it is trying to achieve, I think that is a good compromise.

That said, another approach is to figure out the most readable/maintainable thing, and then go back to add those bits to ensure better interop.

And one last thought: it should be possible to write tests cases that work on master, but fail in this case. That could help us know when the problem is fixed.

gr2m commented:
I’ve to log off for today, just want to share what I’m looking into right now. 

I haven’t fully tested it yet, but I think if we add sth like this to the footer before we return the classes that others inherit from, we should be fine:

```js
const inherits = require('util').inherits
const EventEmitter = require('events').EventEmitter
class Adapter extends EventEmitter {
  // implementation of Adapter
}

function AdapterCompatibleWithCoffeeScript () {
  return new (Function.prototype.bind.apply(Adapter, [ null ].concat([].slice.call(arguments))))
}
inherits(AdapterCompatibleWithCoffeeScript, Adapter)

module.exports = AdapterCompatibleWithCoffeeScript
```

It’s not very pretty, but it _might_ be a good compromise for the time being, so we can move forward with JavaScript implementation while remaining compatibility with existing adapters.

gr2m committed:
add checklist for updating docs in officially supported projects

ping @bkeepers re: https://github.com/hubotio/evolution/pull/4#discussion_r119362201

gr2m committed:
todo to add note for Debian/Ubuntu users

ping @mose re: https://github.com/hubotio/evolution/pull/4#discussion_r119400682

gr2m committed:
_drafts/

link to Hubot 3 milestone

gr2m committed:
_drafts/

clarification of detailed process

ping @technicalpickles re: https://github.com/hubotio/evolution/pull/4#discussion_r119499855

gr2m committed:
typo

gr2m committed:
_drafts/

add list for popular community projects

gr2m committed:
typo

gr2m commented:
Unfortunately I run into a new problem with the [CoffeeScript-compatible export of "classes" mentioned above](https://github.com/hubotio/evolution/pull/4#issuecomment-305557470)

While CoffeeScript can use `AdapterCompatibleWithCoffeeScript` for own extensions, JavaScript cannot with the `class` keyword. So this become tricky, in a nutshell we can have either CoffeeScript’s `class` or JavaScript’s `class` compatibility – not both :(

We could move the logic to export the classes in a CoffeeScript-compatible way into `/index.js` so that the built-in adapters `shell` and `campfire` could still extend the `adapter` class while external modules would get the CoffeeScript-compatible versions when doing `require('hubot')`. But that also means that if someone wants to build their scripts in es2015, then it won’t work unless they directly require the respective class, e.g. `require('hubot/src/adapter')`, but I wouldn't officially support that, paths like `hubot/src/adapter` are internal implementation details and we should be able to change them if needed without breaking the eco system. (that being said, [hubot-mock-adapter does exactly that](https://github.com/blalor/hubot-mock-adapter/blob/5cb366c9150693624050bc2b04e39c83470de5e8/index.js#L23). It would be helpful to find out how many others do the same).

I’d still like to move forward with the CoffeeScript -> JavaScript conversion without breaking existing scripts. What we could do is to add a new file that people can require which gives them es2015-compatible class exports. We can deprecate that in future and remove it eventually. I don’t think it would be too much of an overhead. It would be the exact same logic as `/index.js` without the CoffeeScript-compatibility workaround from my comment above. The only thing we would to do is to update existing scripts that require deep paths like `hubot-mock-adapter` does and change `require('hubot/src/adapter')` to `require('hubot').Adapter`. A code search currently returns [107 results for `"hubot/src/adapter"`](https://github.com/search?utf8=%E2%9C%93&q=%22hubot%2Fsrc%2Fadapter%22&type=Code), people also directly require other classes.

Any thoughts?

See https://github.com/github/hubot/pull/1347/commits/a1e6b1d3e822fa019c4f9cbbef4514d268ef948b for the implementation of the suggestion above

gr2m committed:
_drafts/

address incompatibility of CoffeeScript & es2015 `class` declarations

gr2m commented:
@hubotio/maintainers I think I addressed all feedback and would suggest we merge in the proposal and move on with the pull request at https://github.com/github/hubot/pull/1347. I would move all TODOs from the proposal to the pull request description to keep everyone up-to-date with the progress. 

This is blocking a lot of other efforts to make the Hubot project more contributor / maintainer friendly and is also blocking existing and new pull requests, so I’d like to move forward as fast as possible and if possible move things into follow up issues / PRs

gr2m committed:
add instructions to test a 3rd party adapter with the hubot PR

gr2m committed:
state: In Progress -> In Review

bkeepers commented:
> Unfortunately I run into a new problem with the CoffeeScript-compatible export of "classes" mentioned above

Dang, CoffeeScript not support ES2015 classes is a huge bummer. The workarounds are clever, so 👍  if that works. The explanation of the issue and the workaround should probably be documented in the proposal.

> I think I addressed all feedback and would suggest we merge in the proposal and move on with the pull request at github/hubot#1347.

The [process docs](https://github.com/hubotio/evolution/blob/master/CONTRIBUTING.md#review-process) currently say:  "Once a proposal has been implemented, the Pull Request will be merged." I know that feels a little weird to leave a "proposal" open until the implementation is done, but I think that keeps all of the discussions and documentation in one place.

I think it make sense to leave this open until we finish the implementation, including a release and docs, since some of the details may change.

gr2m commented:
> The explanation of the issue and the workaround should probably be documented in the proposal.

done: https://github.com/hubotio/evolution/blob/4/javascript/_drafts/javascript.md#proposed-solution

> I think it make sense to leave this open until we finish the implementation, including a release and docs, since some of the details may change.

My fault, I misunderstood the process, all clear now

gr2m commented:
My PR is done and ready for review: https://github.com/github/hubot/pull/1347

What would be great is if people could help me test my PR against existing adapters and other 3rd party libraries which depend on `hubot` to make sure they all still work. I’ve tested Slack and it alls seems to work fine. I’ve documented the process of testing at https://github.com/github/hubot/pull/1347#test. Thanks 🙏

gr2m committed:
_drafts/

add note about upcoming Debian release with node@4 included

gr2m commented:
pull request was merged: https://github.com/hubotio/evolution/pull/4#issuecomment-305557470. Will now look into the open todos / follow ups

gasolin commented:
Nice to see ES6 porting is moving on! @gr2m you can also take a look on https://github.com/gasolin/webbybot with is a fork from hubot and focus on translate it to es6 syntax. Test cases are transferred first to make sure everything is still works fine in coffescript / es6 port. 

Here's the related blog post
https://blog.gasolin.idv.tw/2016/03/14/How-we-ported-Hubot-from-Coffeescript-to-ES6/

gr2m commented:
> Where can the README/docs for the current Hubot v2 remain accessible

I would suggest we create a `2.x` branch and link https://github.com/github/hubot/blob/v2.x/docs/index.md directly from our documentation so people could still look up the old docs.

gr2m commented:
@technicalpickles brought up problems with `peerDependencies` in existing hubot scripts that we have to address before releasing `hubot@3`: https://github.com/hubotio/hubot/issues/1057

gr2m committed:
add plan for releasing `hubot@3`

gr2m commented:
I’ve added a release plan for `hubot@3.0.0` to address concerns from hubotio/hubot#1057. See the latest version in the proposal at https://github.com/hubotio/evolution/blob/4/javascript/_drafts/javascript.md#release-hubot300

gr2m commented:
Also added another todo:

> Add documentation for developers of adapters and scripts. If they want to use modern es2015 classes, they must require hubot >= 3 in peerDependencies and `require('hubot/es2015')`. If they want to remain compatible with `hubot@2` then they can’t use es2015 classes. We should document that somewhere and explain it very clearly.

technicalpickles commented:
> pull request was merged: #4 (comment). Will now look into the open todos / follow ups

It feels weird to me that the hubot PR was merged before we'd 'accepted' this proposal, when I had requested changes. I was offline the week that a lot of work was done, and this was merged, but I didn't see discussion around 'dismissing' my feedback since it had been addressed.

> The process docs currently say: "Once a proposal has been implemented, the Pull Request will be merged." I know that feels a little weird to leave a "proposal" open until the implementation is done, but I think that keeps all of the discussions and documentation in one place.
>
>I think it make sense to leave this open until we finish the implementation, including a release and docs, since some of the details may change.

I think this is highlighting a problem with how the evolution flow is intended to work. In particular, if I've requested changes to the proposal (via PR), then it's never blocked from 

Some ideas:

- move discussion to an issue (if there's not already) once the plan has been approved
- update the proposal after it's been merged if implementation details 
- make it clear that the proposal is what is determined at time of planning, and details will change
- consider enabling issues on this repo so they can be tracked here, but be pretty strict about what should be left open

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,463 @@
+  - [ ] Convert generator to JavaScript, release new breaking version with `@next` tag
+  - [ ] Test all `@next` releases
+  - [ ] Set `@latest` tag on all modules
+- [ ] Update documentation
```
The documentation should be updated before the release. I want to make sure that the files that are packaged include updated documentation with how to use the Javascript support.

gasolin commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,463 @@
+
+Convert all source files from CoffeeScript to JavaScript with a tool like [decaffeinate](https://github.com/decaffeinate/decaffeinate), see also [github/hubot#1138](https://github.com/github/hubot/issues/1138).
+
+Make sure the existing tests (still written in CoffeeScript) run against the new JavaScript.
```
My previous hubot -> ES6 porting experience indicate `porting tests first` works well (we done all the porting in 3 weeks). Not sure for the reverse route.

timothywellsjr commented:
Hey guys while running  script/server I am getting the error below. I have the hubbermoji.yml file not sure why it cannot find the file? 

Have you seen this error message before? It seems to be looking for a JS file and not the .yml data file. https://github.com/hubotio/hubot/blob/master/src/robot.js#L376   @mistydemeo suggested that the new code path possibly could be having issues loading files that aren’t actually JS modules.

```C02RK221FVH8:hubot-classic timothywellsjr$ script/server [--debug]
No history available
Hubot> [Tue Jun 20 2017 17:22:00 GMT-0400 (EDT)] ERROR Unable to load /Users/timothywellsjr/github/hubot/bin/hubot-classic/scripts/hubbermoji: Error: Cannot find module '/Users/timothywellsjr/github/hubot/bin/hubot-classic/scripts/hubbermoji'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Robot.loadFile (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/robot.js:353:22)
    at fs.readdirSync.sort.map.file (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/robot.js:376:52)
    at Array.map (native)
    at Robot.load (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/robot.js:376:35)
    at Shell.loadScripts (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/bin/hubot.js:115:9)
    at Shell.g (events.js:291:16)
    at emitNone (events.js:86:13)
    at Shell.emit (events.js:185:7)
    at loadHistory (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/adapters/shell.js:47:19)
    at loadHistory (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/adapters/shell.js:119:12)
    at Shell.run (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/adapters/shell.js:40:5)
    at Robot.run (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/src/robot.js:639:18)
    at Object.<anonymous> (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/bin/hubot.js:112:7)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/lib/coffee-script/register.js:45:36)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/bin/hubot:2:3)
    at Object.<anonymous> (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/hubot/bin/hubot:4:4)
    at Module._compile (module.js:570:32)
    at Object.exports.run (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/lib/coffee-script/coffee-script.js:173:23)
    at compileScript (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/lib/coffee-script/command.js:224:29)
    at compilePath (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/lib/coffee-script/command.js:174:14)
    at Object.exports.run (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/lib/coffee-script/command.js:98:20)
    at Object.<anonymous> (/Users/timothywellsjr/github/hubot/bin/hubot-classic/node_modules/coffee-script/bin/coffee:7:41)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

C02RK221FVH8:hubot-classic timothywellsjr$ 
```

mistydemeo commented:
Yeah, in our current bot this data file lives alongside coffeescript and javascript scripts. Does the new loader try to load anything that's in there, not just js/coffee files?

technicalpickles commented:
@timothywellsjr @mistydemeo I filed https://github.com/hubotio/hubot/issues/1355 to work through that. It does seem look like a regression though.

timothywellsjr commented:
Thanks @technicalpickles is there a workaround to continue testing or will I need to wait for that to be resolved?

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,463 @@
+
+### Update package.json
+
+Remove all CoffeeScript related dependencies and tooling. Update [engines](https://docs.npmjs.com/files/package.json#engines)
```
This step still needs to be done. I see `coffee-script` in there, and also the 0.8.x nodejs engine.

gr2m replied:
startet PR for engines:
https://github.com/hubotio/hubot/pull/1358

We still have the`coffee-script` dependency because [`bin/hubot` still depends on it](https://github.com/hubotio/hubot/blob/d53e399fa0a6a2f0b441aafb869d1640d12785e0/bin/hubot#L1). We left it in so that existing apps written in CoffeeScript will continue to work.

Is there a better way to achieve the same?

technicalpickles commented on _drafts/javascript.md:
```diff
@@ -0,0 +1,463 @@
+
+#### Publish `hubot@3.0.0` with `--tag=next`
+
+- Setup semantic-versioning for `hubot` as described in [`semantic-release` proposal](https://github.com/hubotio/evolution/pull/6).
```
https://github.com/hubotio/evolution/pull/6#issuecomment-308113343 talks about doing one last manual version. If that's the case, this should be removed.

gr2m replied:
Sorry for the confusion, I changed my mind here, I think we can use semantic-release for version 3.0.0 already as we will release with the `@next` tag at first

technicalpickles reviewed (changes requested):
@gr2m asked for my feedback around the release of 3.0, so giving this some 👀 

It looks good overall. Some questions (with line comments):

- are we going to use semantic-release for the initial 3.0 or not?
- package.json still has coffee-script dependency, and old engines sepcified

Not a specific line item, and I may be missing it somewhere else, but before this is done, we'd also need some way to easily access old, existing documentation that is written for coffeescript.

gr2m commented:
### update

- [x] add an automated end-to-end test for hubot scripts written in CoffeeScript (https://github.com/hubotio/hubot/pull/1360)
- [x] add tests for `require('hubot/es2015.js')` (https://github.com/hubotio/hubot/pull/1359)

gr2m commented:
Question: shall we remove loading scripts from `hubot-scripts.json` as it deprecated in 2.x and announced to be removed in 3.x? https://github.com/hubotio/hubot/blob/d53e399fa0a6a2f0b441aafb869d1640d12785e0/bin/hubot.js#L130-L192

gr2m commented:
> Not a specific line item, and I may be missing it somewhere else, but before this is done, we'd also need some way to easily access old, existing documentation that is written for coffeescript.

@technicalpickles it’s one of the open todos in the PR description above

> Where can the README/docs for the current Hubot v2 remain accessible? See https://github.com/hubotio/evolution/pull/4#discussion_r119500396

The simplest would be to create a `2.x` branch and to link to https://github.com/hubotio/hubot/tree/master/docs. I don’t know how we currently deploy to ttps://hubot.github.com/docs/, if we could somehow deploy the docs of a `2.x` branch to https://hubot.github.com/v2/docs/ that’d be great

technicalpickles commented:
> shall we remove loading scripts from hubot-scripts.json as it deprecated in 2.x and announced to be removed in 3.x? 

I'd be 👍 to that. We'd have to make sure to update the generator too https://github.com/hubotio/generator-hubot/pull/91

> I don’t know how we currently deploy to https://hubot.github.com/docs/

It's powered by GitHub pages, coming from a private repo. This repository is included as a submodule under docs, so it generates the right thing. It'd be possible to apply that same pattern to another path.

gr2m committed:
update links to PRs updating peerDependencies

gr2m committed:
update list of peerDependencies PRs & add list for patch releases of scripts

gr2m committed:
remove deprecated scripts from the default install using yeoman generator

gr2m committed:
added video tutorial to convert CoffeeScript to JavaScript

gr2m committed:
allow `0.0.0-development` as hubot version

gr2m committed:
update initial PR status for `hubotio/hubot-redis-brain`

gr2m committed:
✓ Update hubot’s version range `peerDependencies` of default scripts

gr2m committed:
✓ release patch versions for standard scripts

gr2m committed:
add status for `hubot-maps` script

gr2m committed:
✓ hubot-maps patch release with peerDependencies update

gr2m committed:
✓ hubot@next release

gr2m committed:
add link to PR for JavaScript conversion of hubot-rules

technicalpickles commented:
@gr2m I just tried out `npm install --save hubot@next` on a bot I built last week, where I was using hubot-test-helper:

```
npm WARN unmet dependency /Users/technicalpickles/src/parrotbot/node_modules/hubot-test-helper requires hubot@'>= 2.6.0 < 3' but will load
```

I recall that there were some issues with hubot-test-helper related to inheritence and how it uses classes. I think this might be common enough of a module to make sure it works with hubot 3 & javascript.

gr2m commented:
I’m removing `hubot-test-helper` right now from the yeoman script generator. We don’t use it in our scripts any more, the tests with it have been failing by default.

Do you have an app which is using it right now with passing tests?

gr2m committed:
updating progress for hubot generator

gr2m committed:
update laste node@8 stable version for testing

gr2m committed:
_drafts/

complete list of removed default scripts installed by generator

gr2m committed:
_drafts/

all work is done, now on to testing: `npm install --global generator-hubot@next`

technicalpickles commented:
I don't, but could get it running in hubot-for-hubot pretty easily.

gr2m commented:
from my experience, it’s better to keep tests less "magical" and more verbose. The `hubot-test-helper` is helpful when you setup test and you know exactly what you are doing, but for people new to the project they would need to check it out as well. The way we did it e.g. at https://github.com/hubotio/hubot-rules/blob/master/test/rules-test.js has more code, but also is clear simpler to follow for a new contributor, I think

technicalpickles commented:
I think I missed that we were using the hubot-mock-adapter before now. I see https://github.com/blalor/hubot-mock-adapter/pull/7 that is being used there, and that https://www.npmjs.com/package/hubot-mock-adapter-v3 was released using that. Would you mind moving the fork of hubot-mock-adapter to the hubotio org?

There is documentation that recommends using hubot-test-helper, so we'll want to update that too: https://github.com/hubotio/hubot/blob/master/docs/scripting.md#testing-hubot-scripts

gr2m committed:
move documentation todos from issue

technicalpickles commented:
@gr2m was asking for folks to [test the new release](https://github.com/hubotio/evolution/blob/4/javascript/_drafts/javascript.md#test-all-next-releases). I just ran through it without any problem:

```
npm install -g generator-hubot@next
mkdir hubot-next
yo hubot
bin/hubot
```

I tested a few scripts I knew existed (ping, help, the rules), and they responded.

Only thing of note I saw was a message the first time launching `bin/hubot` that a package-lock.json was created and should be committed. That is going to happen for newer npm releases, but that is a general improvement for the generator, and not really specific to the Javascript evolution.

gr2m commented:
> Only thing of note I saw was a message the first time launching bin/hubot that a package-lock.json was created and should be committed. That is going to happen for newer npm releases, but that is a general improvement for the generator, and not really specific to the Javascript evolution.

This will be fixed once we remove the [bin/hubot](https://github.com/hubotio/generator-hubot/blob/master/generators/app/templates/bin/hubot) from the generator and replace it with a `npm start` which just runs the hubot binary provide by the `hubot` package

joeyguerra commented:
Ping.

gr2m commented:
I’m no longer working on Hubot I’m afraid. I cannot help I’m afraid

joeyguerra commented:
Can I maintain it?
